src/main/scala/io/iohk/ethereum/network/discovery/Secp256k1SigAlg.scala

```diff
@@ -0,0 +1,180 @@
+
+  // A normal Secp256k1 signature consists of 2 bigints `r` and `s` followed by a recovery ID `v`,
+  // but it can be just 64 bytes if that's omitted, like in the ENR.
+  override val SignatureBytesSize = 65
+
+  val SignatureWithoutRecoveryBytesSize = 64
+  val PublicKeyCompressedBytesSize = 33
+
+  override def newKeyPair: (PublicKey, PrivateKey) = {
+    val keyPair = crypto.generateKeyPair(secureRandom)
+    val (privateKeyBytes, publicKeyBytes) = crypto.keyPairToByteArrays(keyPair)
+
+    val publicKey = toPublicKey(publicKeyBytes)
+    val privateKey = toPrivateKey(privateKeyBytes)
+
+    publicKey -> privateKey
+  }
+
+  override def sign(privateKey: PrivateKey, data: BitVector): Signature = {
+    val message = crypto.kec256(data.toByteArray)
+    val keyPair = crypto.keyPairFromPrvKey(privateKey.toByteArray)
+    val sig = ECDSASignature.sign(message, keyPair)
+    toSignature(sig)
+  }
+
+  // ENR wants the signature without recovery ID, just 64 bytes.
+  // The Packet on the other hand has the full 65 bytes.
+  override def removeRecoveryId(signature: Signature): Signature = {
+    signature.size / 8 match {
+      case SignatureBytesSize =>
+        Signature(signature.dropRight(8))
+      case SignatureWithoutRecoveryBytesSize =>
+        signature
+      case other =>
+        throw new IllegalArgumentException(s"Unexpected signature size: $other bytes")
+    }
+  }
+
+  override def compressPublicKey(publicKey: PublicKey): PublicKey = {
+    publicKey.size / 8 match {
+      case PublicKeyBytesSize =>
+        // This is a public key without the prefix, it consists of an x and y bigint.
+        // To compress we drop y, and the first byte becomes 02 for even values of y and 03 for odd values.
+        val (xbs, ybs) = publicKey.splitAt(publicKey.length / 2)
+        val y = BigInt(1, ybs.toByteArray)
+        val prefix: Byte =
+          if (y.mod(2) == 0) ECDSASignature.CompressedEvenIndicator
+          else ECDSASignature.CompressedOddIndicator
+        val compressed = PublicKey(BitVector(prefix) ++ xbs)
+        assert(compressed.size == PublicKeyCompressedBytesSize * 8)
+        compressed
+
+      case PublicKeyCompressedBytesSize =>
+        publicKey
+
+      case other =>
+        throw new IllegalArgumentException(s"Unexpected uncompressed public key size: $other bytes")
+    }
+  }
+
+  // The public key points lie on the curve `y^2 = x^3 + 7`.
+  // In the compressed form we have x and a prefix telling us whether y is even or odd.
+  // https://bitcoin.stackexchange.com/questions/86234/how-to-uncompress-a-public-key
+  // https://bitcoin.stackexchange.com/questions/44024/get-uncompressed-public-key-from-compressed-form
+  def decompressPublicKey(publicKey: PublicKey): PublicKey = {
+    publicKey.size / 8 match {
+      case PublicKeyBytesSize =>
+        publicKey
+
+      case PublicKeyCompressedBytesSize =>
+        val point = crypto.curve.getCurve.decodePoint(publicKey.toByteArray)
```

Comment:
my guess that it can also be done by bouncy castle::

```
    val params = new ECPublicKeyParameters(
      curve.getCurve.decodePoint(publicKey.toByteArray),
      curve
    )
    params.getQ.getEncoded(false)
```

we may only need to drop first byte,  as it returns key with indicator pre pended

Reply:
Done, thanks. Shame the Java methods don't take named arguments.